File: Week 5/2.MethodOverloading.java
/*
 * WEEK 5 - FUNCTIONS & RECURSION
 * Topic: Method Overloading
 *
 * DEFINITION:
 * Method overloading = multiple methods with the SAME name but DIFFERENT
 * parameter lists (different number, type, or order of parameters).
 *
 * The compiler picks the right version at compile time — this is called
 * "compile-time polymorphism" or "static dispatch."
 *
 * RULES:
 * - Different parameter count OR different parameter types
 * - Return type alone does NOT distinguish overloaded methods
 *
 * WHY IT MATTERS FOR DSA:
 * Many utility methods (e.g., print, compare, swap) benefit from overloading
 * to handle different data types without renaming.
 *
 * Time Complexity: O(1) per call
 */

public class MethodOverloading {

    // Version 1: two ints
    static int add(int a, int b) {
        System.out.println("add(int, int) called");
        return a + b;
    }

    // Version 2: three ints
    static int add(int a, int b, int c) {
        System.out.println("add(int, int, int) called");
        return a + b + c;
    }

    // Version 3: two doubles
    static double add(double a, double b) {
        System.out.println("add(double, double) called");
        return a + b;
    }

    // Overloaded print utility — prints an array of any primitive type
    static void printArray(int[] arr) {
        System.out.print("int[]: ");
        for (int x : arr) System.out.print(x + " ");
        System.out.println();
    }

    static void printArray(double[] arr) {
        System.out.print("double[]: ");
        for (double x : arr) System.out.print(x + " ");
        System.out.println();
    }

    static void printArray(char[] arr) {
        System.out.print("char[]: ");
        for (char c : arr) System.out.print(c + " ");
        System.out.println();
    }

    public static void main(String[] args) {
        System.out.println(add(2, 3));          // calls version 1
        System.out.println(add(1, 2, 3));       // calls version 2
        System.out.println(add(1.5, 2.5));      // calls version 3

        printArray(new int[]{1, 2, 3});
        printArray(new double[]{1.1, 2.2, 3.3});
        printArray(new char[]{'a', 'b', 'c'});
    }
}

File: Week 5/rust/2.method_overloading.rs
/*
 * WEEK 5 - RUST FUNCTIONS & RECURSION
 * Topic: Method Overloading
 * File: 2.method_overloading.rs
 *
 * CONCEPT:
 *  RUST DOES NOT HAVE TRADITIONAL OVERLOADING. To get similar behaviour,
 *  use one of:
 *   - Generics + traits (compile-time, like C++ templates).
 *   - Multiple methods with different names (idiomatic in Rust).
 *   - Enum dispatch when the variants are known.
 *
 * KEY POINTS:
 *  - `fn add<T: Add>(a: T, b: T) -> T::Output { a + b }` -- generic.
 *  - Functions with the SAME name in the same scope CONFLICT (compile error).
 *  - Trait methods can be implemented per type, giving polymorphism.
 */

use std::ops::Add;
use std::fmt::Display;

// Generic function -- one definition for many types
fn add_generic<T: Add<Output = T> + Display + Copy>(a: T, b: T) -> T {
    let r = a + b;
    println!("add_generic: {a} + {b} = {r}");
    r
}

// Distinct names for distinct overloads
fn add_three(a: i32, b: i32, c: i32) -> i32 { a + b + c }

// Print collection (generic)
fn print_array<T: Display>(label: &str, arr: &[T]) {
    print!("{label}: ");
    for x in arr { print!("{x} "); }
    println!();
}

fn main() {
    let _ = add_generic(2, 3);            // i32
    let _ = add_generic(1.5, 2.5);         // f64
    let _ = add_generic(10_u8, 20);        // u8 (after default-int rule)

    println!("add_three(1,2,3) = {}", add_three(1, 2, 3));

    print_array("ints  ", &[1, 2, 3]);
    print_array("floats", &[1.1, 2.2, 3.3]);
    print_array("chars ", &['a', 'b', 'c']);
    print_array("strs  ", &["hi", "ho"]);
}

/*
 * NOTES:
 *  - Rust prefers GENERICS + TRAITS over name overloading -- one function works
 *    for many types as long as the trait bounds are met.
 *  - For runtime dispatch, use `dyn Trait` (analogous to Java interfaces).
 *  - The orphan rule restricts where you can implement traits across crates.
 */
